### PR TITLE
make Distributed independent of indices starting at 1

### PR DESCRIPTION
**Pull request by oheil**
Proposal to solve issue https://github.com/JuliaLang/julia/issues/34870

### CODE REVIEW

File: stdlib/Distributed/src/macros.jl [hunk 1]
```diff
@@ -260,9 +260,27 @@ function splitrange(N::Int, np::Int)
...
+        chunks[i] = lo:hi
+        lo = hi+1
+    end
+    return chunks
+end
+
 function preduce(reducer, f, R)
-    N = length(R)
-    chunks = splitrange(Int(N), nworkers())
+    chunks = splitrange(firstindex(R),lastindex(R), nworkers())
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/macros.jl]:
```suggestion
    chunks = splitrange(Int(firstindex(R)), Int(lastindex(R)), nworkers())
```

Please also generally use `, ` (comma and space) consistently between arguments.

**Comment by JeffBezanson** [comment]
Thanks! Seems good to me.

File: stdlib/Distributed/src/macros.jl [hunk 2]
```diff
@@ -260,9 +260,27 @@ function splitrange(N::Int, np::Int)
     return chunks
 end
 
+# Statically split range [firstIndex,lastIndex] into equal sized chunks for np processors
+function splitrange(firstIndex::Int, lastIndex::Int, np::Int)
+    each = div(lastIndex-firstIndex+1,np)
+    extras = rem(lastIndex-firstIndex+1,np)
```
Comment (timholy) [inline stdlib/Distributed/src/macros.jl]:
```suggestion
    each, extras = divrem(lastIndex-firstIndex+1, np)
```

**Review by timholy — COMMENTED** [review]
Looks good

**Comment by oheil** [comment]
changed as suggested

**Comment by timholy** [comment]
Thanks. I just realized this doesn't have a test, could you add one? There's a mock implementation of OffsetArrays, see https://github.com/JuliaLang/julia/blob/c2cd601f01849aca85c843f3eb277a0342fee0a6/test/offsetarray.jl#L3-L4

**Comment by oheil** [comment]
Yes, but it will take a few days.

The original function splitrange(N,np) is still in the code. I will check if it used somewhere and if not, should I remove it? What do you think?
Or it is better to mark it debrecated? Can you give me hint on how to deprecate code properly?

**Comment by timholy** [comment]
If it's not used and not exported, I suspect you can just delete it; it doesn't seem to be a likely callee by downstream packages.

If you do decide to deprecate it, try something like
```julia
@deprecate splitrange(N::Int, np::Int) splitrange(1, N, np)
```
If you want the warning to be more informative about why `1` is required, you could mimic https://github.com/JuliaLang/julia/blob/c2cd601f01849aca85c843f3eb277a0342fee0a6/base/deprecated.jl#L219-L225. (Of course this is a function rather than a macro, the only line to pay attention to is the call to `Base.depwarn`.)

**Comment by oheil** [comment]
Removed old version of splitrange and added minimal tests for new splitrange

**Comment by oheil** [comment]
Done. Ready for review.

**Pull request merged**